File: go/experiment.go
package main

type History struct {
	states  [][]float64
	actions []float64
	rewards []float64
}

func NewHistory(stepsNum int) *History {
	return &History{
		states:  make([][]float64, stepsNum),
		actions: make([]float64, stepsNum),
		rewards: make([]float64, stepsNum),
	}
}

type Experiment struct {
	agent                 *Agent
	env                   *Env
	episodesNum, stepsNum int
}

func NewExperiment() *Experiment {
	agent := NewAgent()
	env := NewEnv()
	return &Experiment{
		agent:       agent,
		env:         env,
		episodesNum: 20000000,
		stepsNum:    env.fps * 10,
	}
}

// Run は実験して returns を返す
func (e *Experiment) Run() []float64 {
	returns := make([]float64, e.episodesNum)

	for episode := 0; episode < e.episodesNum; episode++ {
		e.env.ResetEnv()
		hist := e.oneEpisode()
		for _, r := range hist.rewards {
			returns[episode] += r
		}
	}

	return returns
}

func (e *Experiment) Test() *History {
	e.env.ResetEnv()
	e.agent.SetTestParams()
	return e.oneEpisode()
}

func (e *Experiment) oneEpisode() *History {
	hist := NewHistory(e.stepsNum)

	s := e.env.State()
	a := 0.0
	r := 0.0

	for step := 0; step < e.stepsNum; step++ {
		hist.states[step] = s
		hist.actions[step] = a
		hist.rewards[step] = r

		a = e.agent.Action(s)
		e.env.Step(a)
		sNext := e.env.State()
		r = e.env.Reward()
		e.agent.Learn(s, a, r, sNext)

		s = sNext
	}

	return hist
}

File: ruby/experiment.rb
require_relative "agent"
require_relative "env"


class History
  attr_accessor :states, :actions, :rewards

  def initialize(steps_num)
    @states = Array.new(steps_num)
    @actions = Array.new(steps_num)
    @rewards = Array.new(steps_num)
  end
end


class Experiment
  def initialize
    @agent = Agent.new()
    @env = Env.new()

    @episodes_num = 10000000
    @steps_num = @env.fps * 10
  end

  def run
    returns = Array.new(@episodes_num)

    (0...@episodes_num).each do |episode|
      hist = one_episode
      returns[episode] = hist.rewards.sum
    end

    returns
  end

  def test
    @agent.set_test_params
    one_episode
  end

  def one_episode
    hist = History.new(@steps_num)

    @env.reset_state
    s = @env.state
    a = 0.0
    r = 0.0

    (0...@steps_num).each do |step|
      hist.states[step] = s
      hist.actions[step] = a
      hist.rewards[step] = r

      a = @agent.action(s)
      @env.step(a)
      s_next = @env.state
      r = @env.reward
      @agent.learn(s, a, r, s_next)

      s = s_next
    end

    hist
  end
end